servers/physics_2d/broad_phase_2d_bvh.cpp

```diff
@@ -33,7 +33,9 @@
 #include "core/project_settings.h"
 
 BroadPhase2DSW::ID BroadPhase2DBVH::create(CollisionObject2DSW *p_object, int p_subindex, const Rect2 &p_aabb, bool p_static) {
-	ID oid = bvh.create(p_object, true, p_aabb, p_subindex, !p_static, 1 << p_object->get_type(), p_static ? 0 : 0xFFFFF); // Pair everything, don't care?
+	uint32_t tree_id = p_static ? 0 : 1;
+	uint32_t tree_collision_mask = 3;
```

Comment:
Ah it didn't occur to me that there may actually be an overhead in the physics from having e.g. static objects paired. As far as the BVH is concerned, there is no overhead as they aren't moving.

But yes this does lend weight to the idea of having 3 trees, and having areas collide with statics, statics with areas, but no area - area or static - static pairs. We can do this pretty easily now it is all user definable client side.

Yes you are right it would be better to put the interaction changes in a separate PR. :+1:

Reply:
> Is there a physics overhead?

There should be no overhead on the physics side as long as it doesn't actually create new pairs (which should be the case with the cull check).

What I'm wondering is if we're going to have an overhead in the BVH itself if, let's say, you add lots of static bodies all at once in the broadphase. Even if they don't move later, they might cause an initial pairing check with each other.

It seems to me that the closest to the previous behavior in term of performance and functionality would be to:
-Start with using tree masks (but still with just 2 trees, static and non-static)
-Remove the object type thing (it wasn't really used anyway since objects were either paired with everything or nothing)
-Just return true in the cull check

And then in a separate PR, we can add one more tree and fix the area vs. static scenario.

I would be less worried for performance in some corner case scenarios this way.
Does that make sense to you?

Reply:
The logic is now as suggested above (I've used an enum which isn't really necessary, but might help for readability).